### ISSUE TITLE
Making passing scripts into examples

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#137]
re the discussion about making more examples/templates for users. This seems like a great idea.
I had a look at the example suggested in the meeting: `/inst/models/passing/exoPredWLS.R`

Added a stack of comments and explanation now commited.

Some residual thoughts: 

1. Can we change the names of these template-type files to something more discoverable? This script handles a joint continuous+ordinal factor model with definition variables implemented via the 0-variance latent hack, and implements these in RAM and LISREL. The file name `exoPredWLS.R`, however is not discoverable.
2. This and other scripts appear to read the file twice, but neither version of filepath is valid for a user with  `OpenMx` installed. Can we implement a single read which works for users as well as the test-bed? Something like a try  statement that reads from models, or from data, or just uses load? A user stumbling onto this file will likely just give up, assuming it's a broken script.
3. In the RAM script, we use var names found in the data as names for latents. Is that legit? 
4. Why do we fix the variance 	for the ordinal variables at .5?
5. Is "slopes" the same as means (will make people think of multi-level. I couldn't see what the definition variables are supposed to be doing. Some comments about the model structure would help here.
6. 	There's a hack setting the thresholds matrix: `jr$expectation$thresholds <- 'T' ` Can that be done in a more conventional/documented fashion, i.e. `mxExpectationRAM(M = NA, dimnames = NA, thresholds = "T", threshnames = ???)`
7. There's a section on testing "permutation" code. Unclear in what sense this is a permutation... Doesn't seem to test anything as a result? Should this be "iterate over model types allowed by the buildModel function"

**Comment by jpritikin** [comment]
First of all, thanks for your comments. I think this is really helpful.

> This and other scripts appear to read the file twice, but neither version of filepath is valid for a user with  `OpenMx` installed. Can we implement a single read which works for users as well as the test-bed? Something like a try  statement that reads from models, or from data, or just uses load? A user stumbling onto this file will likely just give up, assuming it's a broken script.

Yes, that's a great idea. Once we figure out the best fix, I'd be happy to go through all the tests and apply changes so everything is consistent.

> In the RAM script, we use var names found in the data as names for latents. Is that legit?

Yes, it is legit and helps reduce cognitive load.

> Why do we fix the variance for the ordinal variables at .5?

That's measurement error. I don't recall why I picked 0.5. It's arbitrary.

> Is "slopes" the same as means (will make people think of multi-level. I couldn't see what the definition variables are supposed to be doing. Some comments about the model structure would help here.

The `slopes` terminology was copied for lavaan's internal source code. In retrospect, I don't think it's a particularly evocative term. We can better document it as an `exogenous predictor regression coefficient`.

> There's a hack setting the thresholds matrix: `jr$expectation$thresholds <- 'T' ` Can that be done in a more conventional/documented fashion, i.e. `mxExpectationRAM(M = NA, dimnames = NA, thresholds = "T", threshnames = ???)`

Yes, but we already have `mxModel(type='RAM', ...)` so `mxExpectationRAM` is redundant except for the need to set the name of the thresholds matrix. I'm not sure what is the most concise solution. Is it feasible to provide a default name for the threshold matrix whether or not there are ordinal indicators?

> There's a section on testing "permutation" code. Unclear in what sense this is a permutation... Doesn't seem to test anything as a result? Should this be "iterate over model types allowed by the buildModel function"

It's a great idea to turn tests into demos, but this section of code probably looks nonsensical as a demonstration. The point of that code is to exercise some tricky code in the backend. So it's important for testing, but not useful as a demo. As part of this documentation effort, it probably makes sense to split up files into demonstration and truly testing, non-demo parts.

**Comment by RMKirkpatrick** [comment]
FWIW, I'd like to point out a script I wrote, [/models/passing/GeneralizedEstimatingEquations.R](https://github.com/OpenMx/OpenMx/blob/master/inst/models/passing/GeneralizedEstimatingEquations.R), as an example of a test script that is also a readable example script.

**Comment by tbates** [comment]
re using a var name as a latent name: Is that just the special case of when the user declares the variable a definition variable, then that name becomes available to be used as a latent name as well?

re: pathnames: Some sort of mxLoad() function could be smart about detecting the environment and parsing the pathname appropriately, or trying several?

re: thresh: What would the impact be of a default thresh="Thresholds" be? I could imagine this going awry if the model looked up and found it in the environment? or if it's the trigger to treat things as factors?

"exogenous predictor regression coefficient" might be more odd than slopes :-) just a sentence about how the model works would be good.

**Comment by tbates** [comment]
@RMKirkpatrick perhaps we could start moving files into `inst/model/demonstration` if they meet some criterion for commenting and explanation?

**Comment by RMKirkpatrick** [comment]
@tbates I don't know about "moving" them, if that means taking them out of the test suite.  Symlinking them, maybe?

**Comment by jpritikin** [comment]
> This and other scripts appear to read the file twice, but neither version of filepath is valid for a user with  `OpenMx` installed. Can we implement a single read which works for users as well as the test-bed? Something like a try  statement that reads from models, or from data, or just uses load? A user stumbling onto this file will likely just give up, assuming it's a broken script.

Probably the correct fix is to package these datasets properly in the data subdirectory like demoOneFactor.

**Comment by tbates** [comment]
@jpritikin yes - `data` already contains a lot of redundant files though (simple variations on a theme, and no documentation)

So we should probably also see whether there are files already in data that can be used in the tests.

**Comment by tbates** [comment]
@RMKirkpatrick idea would be to create a folder like passing, still in the test suite, which contains files that are well-enough documented to be useful templates for users.

**Comment by jpritikin** [comment]
> idea would be to create a folder like passing, still in the test suite, which contains files that are well-enough documented to be useful templates for users.

We already have a demo folder. Why not use that?

**Comment by tbates** [comment]
Sounds good: is `demo` checked in the test suite? If, yes, then let's each of us move (or, more likely copy cut down user-focussed instead of testing-focussed versions) files there from models/passing as we document them well enough to be used as templates for users.

Let's also create a web page which consists of links to the demo folder files, with brief explanations.

**Comment by mcneale** [comment]
Agreed, seems good path forward.

Cheers
Mike

Michael C. Neale, Ph.D.
Professor, Virginia Institute for Psychiatric and Behavioral Genetics
Virginia Commonwealth University
Shipping Address: 800 East Leigh St. Suite 1-156, Richmond VA 23219-1534
Mailing Address : Box 980126 Richmond VA 23298-0126
Telephone: (804) 828-3369

On Feb 19, 2019, at 10:37 AM, Tim Bates <notifications@github.com<mailto:notifications@github.com>> wrote:


Sounds good: is demo checked in the test suite? If, yes, then let's each of us move (or, more likely copy cut down user-focussed instead of testing-focussed versions) files there from models/passing as we document them well enough to be used as templates for users.

Let's also create a web page which consists of links to the demo folder files, with brief explanations.

**Comment by tbates** [comment]
I changed a couple of files from 
```R
jointData <- suppressWarnings(try(read.table("models/passing/data/jointdata.txt", header=TRUE), silent=TRUE))
jointData <- read.table("data/jointdata.txt", header=TRUE)
```
to 
```R
load("jointdata")
```

see how this works out

**Comment by jpritikin** [comment]
> So we should probably also see whether there are files already in data that can be used in the tests.

@tbates Usually the data need a specific structure. However, I hate to add more undocumented data. It seems like the best way forward is to generate the data on the fly, but it takes a bit of work to simulate data with a suitable distribution.

**Comment by jpritikin** [comment]
Today I learned that `suppressWarnings(RNGversion("3.5"))` can ensure that randomly generated data remain identical across R versions.
